flowmachine/flowmachine/core/cache.py

```diff
@@ -0,0 +1,384 @@
+    ----------
+    connection : Connection
+    query_id : str
+        md5 id of the query
+
+    Returns
+    -------
+    Query
+        The original query object.
+
+    """
+    qry = f"SELECT obj FROM cache.cached WHERE query_id='{query_id}'"
+    try:
+        obj = connection.fetch(qry)[0][0]
+        return pickle.loads(obj)
+    except IndexError:
+        raise ValueError(f"Query id '{query_id}' is not in cache on this connection.")
+
+
+def get_cached_queries_by_score(connection: "Connection") -> List[Tuple["Query", int]]:
+    """
+    Get all cached queries in ascending cache score order.
+
+    Parameters
+    ----------
+    connection : Connection
+
+    Returns
+    -------
+    list of tuples
+        Returns a list of cached Query objects with their on disk sizes
+
+    """
+    qry = """SELECT obj, table_size(tablename, schema) as table_size
+        FROM cache.cached
+        WHERE NOT cached.class='Table'
+        ORDER BY cache_score_multiplier*((compute_time/1000)/table_size(tablename, schema)) ASC
+        """
+    cache_queries = connection.fetch(qry)
+    return [(pickle.loads(obj), table_size) for obj, table_size in cache_queries]
+
+
+def shrink_one(connection: "Connection", dry_run: bool = False) -> "Query":
+    """
+    Remove the lowest scoring cached query from cache and return it and size of it
+    in bytes.
+
+    Parameters
+    ----------
+    connection : "Connection"
+    dry_run : bool, default False
+        Set to true to just report the object that would be removed and not remove it
+
+    Returns
+    -------
+    tuple of "Query", int
+        The "Query" object that was removed from cache and the size of it
+    """
+    obj_to_remove, obj_size = get_cached_queries_by_score(connection)[0]
+
+    logger.info(
+        f"{'Would' if dry_run else 'Will'} remove cache record for {obj_to_remove.md5} of type {obj_to_remove.__class__}"
+    )
+    logger.info(
+        f"Table {obj_to_remove.table_name} ({obj_size} bytes) {'would' if dry_run else 'will'} be removed."
+    )
+
+    if not dry_run:
+        obj_to_remove.invalidate_db_cache(cascade=False, drop=True)
+    return obj_to_remove, obj_size
+
+
+def shrink_below_size(
+    connection: "Connection", size_threshold: int = None, dry_run: bool = False
+) -> "Query":
+    """
+    Remove queries from the cache until it is below a specified size threshold.
+
+    Parameters
+    ----------
+    connection : "Connection"
+    size_threshold : int, default None
+        Optionally override the maximum cache size set in flowdb.
+    dry_run : bool, default False
+        Set to true to just report the objects that would be removed and not remove them
+
+    Returns
+    -------
+    list of "Query"
+        List of the queries that were removed
+    """
+    initial_cache_size = size_of_cache(connection)
+    removed = []
+    logger.info(
+        f"Shrinking cache from {initial_cache_size} to below {size_threshold}{'(dry run)' if dry_run else ''}."
```

Comment:
Missing space before `(dry run)`?
```suggestion
        f"Shrinking cache from {initial_cache_size} to below {size_threshold}{' (dry run)' if dry_run else ''}."
```